yente/search/indexer.py

```diff
@@ -202,3 +214,105 @@ async def update_in_thread() -> None:
     )
     thread.start()
     # asyncio.to_thread(update_index, force=force)
+
+
+class DeltasNotAvailable(Exception):
+    pass
+
+
+async def get_deltas_from_version(
+    version: str, dataset: Dataset
+) -> AsyncGenerator[Dict[str, Any], None]:
+    """
+    Get deltas from a specific version of a dataset.
+    """
+    version = version.replace(settings.INDEX_VERSION, "")
+    try:
+        async for line in load_json_lines(dataset.delta_path(version), version):
+            yield line
+    except HTTPStatusError as exc:
+        if exc.response.status_code == 404:
+            raise DeltasNotAvailable(f"No deltas found for {version}")
+
+
+async def get_next_version(dataset: Dataset, version: str) -> str | None:
+    """
+    Get the next version of a dataset if versions are available.
+    Return None if the dataset is up to date.
+    """
+
+    available_versions = [
+        settings.INDEX_VERSION + v for v in await dataset.available_versions()
+    ]
+    try:
+        ix = available_versions.index(version)
+    except ValueError:
+        raise DeltasNotAvailable(
+            f"Current version of dataset not found in available versions: {dataset.name}, {version}"
+        )
+    if ix == len(available_versions) - 1:
+        log.info(
+            "Dataset is up to date.",
+            dataset=dataset.name,
+            current_version=version,
+        )
+        return None
+    next_version = available_versions[ix + 1]
+    return next_version
+
+
+async def delta_update_index(
+    dataset: Dataset, provider: SearchProvider, force: bool = False
+) -> bool:
+    if not dataset.load:
+        log.debug("Dataset is not going to be loaded", dataset=dataset.name)
+        return False
+    clone = None
+    try:
+        current_version = await get_current_version(dataset, provider)
+        if current_version is None:
+            raise Exception("No index found for dataset.")
+        index = Index(provider, dataset.name, current_version)
+        target_version = await dataset.newest_version()
+        # If delta versioning is not implemented, update the index from scratch.
+        if target_version is None:
+            raise DeltasNotAvailable(f"No versions available for {dataset.name}")
```

Comment:
Is this where it'd branch out into a full re-index?

Reply:
Yep, any exception should cause a full re-index.